concordium-consensus/src/Concordium/KonsensusV1/Types.hs

```diff
@@ -0,0 +1,804 @@
+                then do
+                    Present <$> get
+                else return Absent
+        tmAggregateSignature <- get
+        return TimeoutMessageBody{..}
+
+-- |The 'TimeoutSignatureMessage' associated with a 'TimeoutMessageBody'.
+tmSignatureMessage ::
+    -- |Genesis block hash
+    BlockHash ->
+    TimeoutMessageBody ->
+    TimeoutSignatureMessage
+tmSignatureMessage tsmGenesis tmb =
+    TimeoutSignatureMessage
+        { tsmRound = tmRound tmb,
+          tsmQCRound = qcRound (tmQuorumCertificate tmb),
+          tsmGenesis
+        }
+
+-- |A timeout message including the sender's signature.
+data TimeoutMessage = TimeoutMessage
+    { -- |Body of the timeout message.
+      tmBody :: !TimeoutMessageBody,
+      -- |Signature on the timeout message.
+      tmSignature :: !BlockSig.Signature
+    }
+    deriving (Eq, Show)
+
+instance Serialize TimeoutMessage where
+    put TimeoutMessage{..} = do
+        put tmBody
+        put tmSignature
+    get = do
+        tmBody <- get
+        tmSignature <- get
+        return TimeoutMessage{..}
+
+-- |Byte representation of a 'TimeoutMessageBody' used for signing the timeout message.
+timeoutMessageBodySignatureBytes :: TimeoutMessageBody -> BS.ByteString
+timeoutMessageBodySignatureBytes body = runPut $ do
+    putByteString "TIMEOUTMESSAGE."
+    put body
+
+-- |Sign a timeout message.
+signTimeoutMessage :: TimeoutMessageBody -> BakerSignPrivateKey -> TimeoutMessage
+signTimeoutMessage tmBody privKey = TimeoutMessage{..}
+  where
+    msg = timeoutMessageBodySignatureBytes tmBody
+    tmSignature = BlockSig.sign privKey msg
+
+-- |Check the signature on a timeout message.
+checkTimeoutMessageSignature :: BakerSignVerifyKey -> TimeoutMessage -> Bool
+checkTimeoutMessageSignature pubKey TimeoutMessage{..} =
+    BlockSig.verify pubKey (timeoutMessageBodySignatureBytes tmBody) tmSignature
+
+-- |Projections for the data associated with a baked (i.e. non-genesis) block.
+class BakedBlockData d where
+    -- |Quorum certificate on the parent block.
+    blockQuorumCertificate :: d -> QuorumCertificate
+
+    -- |Parent block hash.
+    blockParent :: d -> BlockHash
+    blockParent = qcBlock . blockQuorumCertificate
+
+    -- |'BakerId' of the baker of the block.
+    blockBaker :: d -> BakerId
+
+    -- |Signature verification key of the baker of the block.
+    blockBakerKey :: d -> BakerSignVerifyKey
+
+    -- |If the previous round timed-out, the timeout certificate for that round.
+    blockTimeoutCertificate :: d -> Optionally TimeoutCertificate
+
+    -- |If this block begins a new epoch, this is the finalization entry that finalizes the
+    -- trigger block.
+    blockEpochFinalizationEntry :: d -> Optionally FinalizationEntry
+
+    -- |The 'BlockNonce' generated by the baker's VRF.
+    blockNonce :: d -> BlockNonce
+
+    -- |The baker's signature on the block.
+    blockSignature :: d -> BlockSignature
+
+instance BakedBlockData Void where
```

Comment:
Maybe document what this instance is for

Reply:
Or remove it? What is the purpose of it other than it's technically possible?

Reply:
Removed.